### PR TITLE
Random: add support for multiple constraints

### PR DESCRIPTION
**Pull request by maartenuni**
### Additions are public 2 functions:
1. uil.randomization.randomizeStimuliConstraints()
This function takes an object that describes the which keys should be present in an "testitem", and also how many times it may the value may be repeated in a row.

2. uil.randomization.stimuliMeetConstraints().
This function is mainly for testing purposes.

### I also added 2 private functions:

1. randomizePrivate(), this is the function that does the hard work. It basically works via the same algorithm as before, it just is able to handle multiple constraints.
2. allowPushItem(), this function checks whether an item can be pushed onto the (re)ordered items without violating the constraints.

### Changes:
I changed the signature of uil.randomisation.randomizeStimuli()
From:
```javascript
context.randomizeStimuli = function(
    original_stimuli,
    max_same_type = 2,
    type_key = 'item_type',
    attempt = 0,
)
```
to
```javascript
context.randomizeStimuli = function(
    original_stimuli,
    max_same_type = 2,
    type_key = 'item_type'
)
```
As the **attempt** parameter shouldn't be touched by callers anyway according it's documentation. Now i moved the attempt parameter into the private function in such way, it isn't exposed to end users but still makes sure that the randomizePrivate() function doesn't recurse into oblivion.

### Unit tests
I also added some "unit" tests in the file browser-random-tests.html.
I was looking into jest (unit test framework for javascript), but I couldn't get it to work with the jspsych-uil-utils
One can see in the developer console whether the test run and if the assertions work out well.

There is one function with hard constraints, that I hope to get to work. It's currently failing, the basic test work out pretty well I think.

### CODE REVIEW

File: jspsych-uil-randomization.js [hunk 1]
```diff
@@ -32,6 +32,113 @@ else
...
+        if (nth_try < 0) {
+            throw new RangeError("nth_try is < 0");
+        }
+
+        if (nth_try >= max_tries) {
+            console.log(
+                "Unable to randomize according the constraints, "       +
+                "Perhaps it is a good idea to make the constraints "    +
+                "less strict."
+            );
```
Comment (tymees) [inline jspsych-uil-randomization.js:114]: I'd say 'console.warning' is more apropiate here. .log is more for informational, but this message is definetly more than informational

Comment (maartenuni) [reply jspsych-uil-randomization.js:114]: Agreed!

Comment (maartenuni) [reply jspsych-uil-randomization.js:114]: I now see, that the original code even used console.error(), that might also be appropriate.

Comment (tymees) [reply jspsych-uil-randomization.js:114]: Yeah when I was thinking it over just now I thought the same thing, as it's a show-stopping problem

File: jspsych-uil-randomization.js [hunk 2]
```diff
@@ -32,6 +32,113 @@ else
...
+                );
+            }
+
+            let selection = randomized.slice(-max);
+            selection.push(item);
+
+            let count = 0;
+            selection.forEach(stimulus => count += stimulus[key] === value);
+            if (count > max)
+                return false;
```
Comment (tymees) [inline jspsych-uil-randomization.js:74]: I think some comments could help here; I was about to write something (stupid) till I finally got what you're doing here. (I though you were checking the whole list of chosen items, but then I realised you were slicing the selection first).

Something like:
before line 68: "Slice the given array such that we have the last `max` number of items"
before line 72: "Check if the amount of item types in the sliced array does not exceed the contstraint"

**Comment by tymees** [comment]
Looks good! It is definitely a more clever and cleaner implementation than my old version. Aside from the two comments I made in the code I have one more comment: I'd rather put the test html file in a separate folder called 'tests' inside the repo. Keeps it from cluttering up the root. (You can import the library using '../')

File: jspsych-uil-randomization.js (hunk 1 above)
Comment (tymees) [reply jspsych-uil-randomization.js]: While I like your due diligence, you don't _have_ to comment the fixing commits in such tiny code reviews like this. (Unless it helps you of course)

Comment (tymees) [reply jspsych-uil-randomization.js]: (Btw, if you're doing it to mark it as 'done', you can also use the 'Resolve conversation' button on the bottom)

File: jspsych-uil-randomization.js (hunk 2 above)
Comment (maartenuni) [reply jspsych-uil-randomization.js]: Add some comments in [6a7ae08](https://github.com/UiL-OTS-labs/jspsych-uil-utils/pull/20/commits/6a7ae08047d2705dcc6a98163f2e8b99f9cde5f4)

**Comment by maartenuni** [comment]
> Maybe in the future we should look into setting something up to automate the testing in GitHub. (Maybe add a Node script?)

Yes something like that would be nice.

File: jspsych-uil-randomization.js (hunk 2 above)
Comment (tymees) [reply jspsych-uil-randomization.js]: That's much clearer, thank you!

**Comment by tymees** [comment]
I'm not sure replacing a O(n) algorithm with a O(n^n) is such a great idea from a performance standpoint.... Is passing the hard test really that important?

**Comment by tymees** [comment]
If so, I'd rather make the more expensive algorithm an option alongside the less-correct but more performant solution

**Comment by maartenuni** [comment]
Ty, would you be so kind to have one look at my latest changes. I changed the algorithm. I think its more robust, definitely not perfect.
Previously we were "fixing" the input by trying random items, the current algorithm tries to fix the order by elaborately finding items that meet the test constraints. Since it tries to find items in an Array of randomized items, the first matching stimulus is returned. I also think that it's faster in most cases.
I see your comment comming in, shall we meet on teams?

**Comment by maartenuni** [comment]
Well benchmarks depend on the browser. On firefox the old version is slower than the new version of the algorithm, on chrome the old version is slightly faster. Both functions seem to behave roughly linear. Add 10 times as many items, and it will take 10 times as much time.
So perhaps use the uil.randomization.randomizeStimuliConstraints() by default, and with unbalanced stimuli you can think about using the uil.randomization.shuffleStimuliConstraints(). I have added this to the comments of the public functions.

**Pull request merged**